Allocating a lot less in ExceptionFormatter.GetFormattedException

Scooletz opened the pull request:
This PR reduces the amount of memory used by `ExceptionFormatter.GetFormattedException` by passing using all the way up/down a single `StringBuilder`. The amount of allocated memory shown by dotMemory have been reduced and for the following tests is following

|  Test name | Before (bytes)  | After (bytes)  |  
|---|---|---|
| FormatException_OutputsExpectedAsyncMethodParameters  | 294383 	  | 271439|
| FormatException_OutputsMethodParameters | 78014			  | 53774|
| FormatException_RemovesAsyncFrames  | 74614			  | 50450|
| FormatException_ResolvesAsyncMethodNames |94052			|69418|

This PR includes several commits that were put at the beginning and reverted at the end. 

The first one is related to `dotMemory` as I failed to find a benchmarking for this. I didn't want to blow it up out proportion so used locally dotMemory for finding the gaps.

The other commit is related to assertions for the formatted exceptions. I found the number of assertions no big enough to make me sure about the outcome, so I captured the initial output and used it through the whole process.

Even if this PR seems flaky, the implementation of `ExceptionFormatter.GetFormattedException` is good to be copied if/when the benchmarking and stronger assertions are added to the suite. Again, I'm not prescriptive, just trying to bring the reasoning and the work behind the PR

Scooletz committed:
temporary memory measurement in ExceptionFormatter tests

Scooletz committed:
dummy full text assertions

Scooletz committed:
to StringBuilder part 1

Scooletz committed:
to StringBuilder part 2

Scooletz committed:
full text assertions removed

Scooletz committed:
removed dotMemory

Scooletz commented:
I think the failure might be related to https://github.com/Azure/azure-webjobs-sdk/pull/2505#issuecomment-638818228

Soft ping for @brettsam  😉

Scooletz committed:
tests back to netcoreapp2.1

Scooletz committed:
explicit name tuple elements to pass the CI build

Scooletz commented:
After fixup with the tuple naming, the build fails with the error as below. Is it possible for any from the team to take a look and validate what is failing in here?

```
dotnet.exe : [xUnit.net 00:06:22.0680998]     Microsoft.Azure.WebJobs.Host.EndToEndTests.SingletonEndToEndTests.SingletonTriggerFunction_MultipleConcurrentInvocations_InvocationsAreSerialized [FAIL]
At C:\projects\azure-webjobs-sdk-rqm4t\runtests.ps1:15 char:3
+   & dotnet $cmd
+   ~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: ([xUnit.net 00:0...rialized [FAIL]:String) [], RemoteException
    + Full
```

Scooletz commented:
@brettsam Sorry for pinging you, but if you or anyone help me with the https://github.com/Azure/azure-webjobs-sdk/pull/2510#issuecomment-696613493 maybe I could move on and finalize this work :-)

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Diagnostics/ExceptionFormatter.cs:
```diff
@@ -26,92 +26,106 @@ public static class ExceptionFormatter
                                   let declaringType = method?.DeclaringType
                                   where ShouldShowFrame(declaringType)
-                                  select FormatFrame(method, declaringType, frame);
+                                  select (method: method, declaringType: declaringType, frame: frame);
```
What if you kept FormatFrame here instead of creating the tuples and then iterating through them?

Scooletz replied:
Oh my 😅  I simplified it by pushing it down to `FormatFrame`

brettsam commented:
It may just be a flaky test -- I've re-run it to see. I also added one more question...

Scooletz committed:
format frame simplified

Scooletz commented:
@brettsam I amended it. And it's 🟢

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Diagnostics/ExceptionFormatter.cs:
```diff
@@ -26,92 +26,108 @@ public static class ExceptionFormatter
+            var method = frame.GetMethod();
+            var declaringType = method?.DeclaringType;
+            
+            if (ShouldShowFrame(declaringType) == false)
```
This is a really nitpicky comment :-) -- but can you make this `if (!ShouldShowFrame(declaringType))`?

Scooletz replied:
Fixed in https://github.com/Azure/azure-webjobs-sdk/pull/2510/commits/4b6d6d71dbaf5894c1f17ac6598173ab5ef0e5fb

Scooletz committed:
proper negation

Scooletz committed:
two more string allocs killed for async

Scooletz commented:
I've fixed the negation style and removed two more stringy allocs in https://github.com/Azure/azure-webjobs-sdk/pull/2510/commits/ccecaae49f97a78da0affc78c895d429181c55e7

I thinks this is it for this PR. I'd propose some kind of caching for all the calculated values that are based on `MethodInfo` and `Type` but this is a separate PR I think. I'll leave this one as an alloc-related only, without caching applied.

Soooo, are we ok with this @brettsam ? 😄

brettsam commented:
Thanks @Scooletz!

Scooletz commented:
👍 @brettsam
